Publish Blueprint

dtaylor113 opened the pull request:
Hi,
This PR integrates SUI and the PublishBlueprint API, as well as showing read-only versions of several screens for published blueprints.  Please see the following document for various screenshots: [Unpublished .vs Published screenshots](https://docs.google.com/a/redhat.com/document/d/1C4K5sTvCza7l3vp2DwRILiLYCjI006kMTr16H4weGTc/edit?usp=sharing)

PR also fixes Blueprint Details Dlg. tab selection (broke after recent ui-bootstrap upgrade), and expand/collapse Advanced Options.

@serenamarie125 

JIRA: [CFUX-43](https://patternfly.atlassian.net/browse/CFUX-43)
Pivotal: [[UI] Publish Blueprint](https://www.pivotaltracker.com/story/show/134730977)

chriskacerguis commented:
@dtaylor113 could you please add the link to the pivotal card?

chriskacerguis commented:
@AllenBW could you please take a look?

chriskacerguis commented:
@dtaylor113 can you please take a look at the CodeClimate issues?

dtaylor113 commented:
@chriskacerguis , added Pivotal link & looking at code climate issues

AllenBW commented:
👀

AllenBW commented on client/app/components/blueprints/blueprint-details-modal/blueprint-details-modal-service.factory.js:
```diff
@@ -81,20 +81,28 @@
       {"title": __('Action Order'), "active": false},
     ];
 
+    if (vm.blueprint.read_only) {
```
Action not required: consider consider using ternary operator to for conciseness 🌮

dtaylor113 replied:
Hi @AllenBW , do you mean use: `vm.activeTab =  vm.blueprint.read_only ? 1 : 0; `

AllenBW replied:
yeah totally!

dtaylor113 replied:
ok, got it :-)

AllenBW commented on client/app/components/blueprints/blueprint-details-modal/blueprint-details-modal-service.factory.js:
```diff
@@ -260,35 +278,11 @@
-        saveFailure();
-
-        return;
+        BlueprintsState.publishBlueprint(vm.blueprint).then(saveSuccess, publicationFailure);
```
Action not required: consider using names `success` and `failure`, collisions avoided with similar functions in a controller by declaring  aforementioned functions in the parent function

AllenBW commented on client/app/components/blueprints/blueprint-details-modal/blueprint-details-modal-service.factory.js:
```diff
@@ -301,18 +295,36 @@
+        EventNotifications.error(__('There was an error publishing blueprint ') + "'" + vm.blueprint.name + "': " + errorMsg);
+      }
+
+      /*
```
eeekkk !? comments ?! 😕 🇫🇰

AllenBW commented on client/app/components/blueprints/blueprint-editor/_blueprint-editor.sass:
```diff
@@ -1,3 +1,6 @@
+.capitalize
```
There exists angular filter uppercase, would recommend use of that vs style class

https://docs.angularjs.org/api/ng/filter/uppercase

dtaylor113 replied:
I'm looking for camel case/initial caps.  Couldn't find an angular filter for that :frowning:

AllenBW replied:
OH true, that doesn't exist (but maybe a filter we should add if this is something that has value elsewhere)

AllenBW commented on client/app/components/blueprints/blueprints-list.directive.js:
```diff
@@ -44,11 +44,16 @@
         blueprint.ui_properties.service_catalog.name = categoryName[0].name;
       }
+
+      blueprint.read_only = false;
```
Action not required: consider ternary here as well

AllenBW commented on client/app/components/blueprints/blueprints-list.directive.js:
```diff
@@ -188,7 +194,13 @@
         }
       } else {
-        BlueprintDetailsModal.showModal('publish', item);
+        BlueprintsState.getTagsForItem('blueprints', item.id).then(function(tags) {
```
Action not required: consider using named function here (rather than nameless)

AllenBW commented on client/app/services/blueprints-state.service.js:
```diff
@@ -78,6 +78,57 @@
+        }, savefailure);
+      } else {
+        // just publish the unchanged blueprint
+        blueprint.publishBlueprintOp(tmpBlueprint).then(function() {
```
Action not required: consider using a named function `success` and subsequent `failure`

(really at the end of the day this removes the thinking from naming convention, something i hate to spend time on "oooooh what should i name *THIS*)

dtaylor113 replied:
Yeah, but it gets a little more difficult with chained promises, can't call every function success and failure.

AllenBW replied:
yeah yeah!! up to you... though maybe i'm using too liberally, most feedback i'm providing is to more align voice of pr with voice of existing codebase

to the second point totally agree!  it doesn't always make sense, but most of time ya can get away with the simple implementations

AllenBW commented on client/app/services/blueprints-state.service.js:
```diff
@@ -78,6 +78,57 @@
+      }
+
+      function savefailure() {
+        $log.info("Failed to save '" + tmpBlueprint.name + "' before publication.");
```
why $log and not EventNotification?

dtaylor113 replied:
The calling parent displays the EventNotification.

dtaylor113 commented:
Hi @AllenBW, thanks for the review!  I addressed most of your comments.  Didn't think it made sense to use named functions in when chaining promises.  Thanks!

simaishi commented:
@dtaylor113 is there a BZ for this?

dtaylor113 commented:
Hi @simaishi , a Bugzilla? Not that I'm aware of.  There are jira and pivotal story links in the description.

simaishi commented:
@dtaylor113 Tried to backport to Euwe, but 18 out of 21 files are not clean cherry-pick. It looks like some of the files moved in https://github.com/ManageIQ/manageiq-ui-service/pull/244, so that might also be contributing to the conflicts...  Please resolve conflict and make Euwe-specific PR (referencing this one) or suggest other PRs to backport.

simaishi commented:
@dtaylor113 @chriskacerguis Just ping'ing about Euwe branch. If this is not needed in Euwe, please remove the euwe labels. Thanks!

chriskacerguis commented:
@simaishi euwe/yes tag removed.
